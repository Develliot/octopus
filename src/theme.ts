import { DefaultTheme } from 'styled-components';

export const fontSizes: { [key: string]: string } = {
    extraTiny: '0.6.25rem',
    tiny: '0.75rem',
    small: '0.875rem',
    medium: '1rem',
    large: '1.125rem',
    extraLarge: '1.25rem',
    extraExtraLarge: '1.5rem',
    extraExtraExtraLarge: '2rem',
};

export const breakpoints = {
    small: '768px',
    large: '1400px',
};

export const globalPadding = '10vw';

export type SpacingType =
    | 'tiny'
    | 'small'
    | 'medium'
    | 'large'
    | 'extraLarge'
    | 'extraExtraLarge';

export const spacing: { [key in SpacingType]: string } = {
    tiny: '0.5rem',
    small: '1rem',
    medium: '1.5rem',
    large: '3rem',
    extraLarge: '4rem',
    extraExtraLarge: '8rem',
};

export type ColorType =
    | 'white'
    | 'lightGrey'
    | 'blueGrey'
    | 'darkBlue'
    | 'blue'
    | 'lightBlue'
    | 'pink';

export const colors: { [key: string]: string } = {
    white: '#ececec',
    lightGrey: '#d4d4d4',
    blueGrey: '#34588c',
    darkBlue: '#040e28',
    blue: '#011739',
    lightBlue: '#5bf8ff',
    pink: '#ff1b7e',
};

const theme: DefaultTheme = {
    fontSizes,
    breakpoints,
    globalPadding,
    spacing,
    colors,
};

export default theme;
