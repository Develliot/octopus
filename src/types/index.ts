export type ProductType = {
    id: number;
    title: string;
    wattage: number;
    packetSize: number;
    price: number;
    description: string;
    imageUrl: string;
    specifications: {
        brand: string;
        weight: string;
        dimensions: string;
        modelNumber: string;
        colour: string;
    };
};
