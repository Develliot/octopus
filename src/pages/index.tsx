import { NextPageContext } from 'next';
// import isoFetch from 'isomorphic-unfetch';

import data from 'src/data/lightBulbs.json';

import { H1, Paragraph } from 'src/components/Typography';
import {
    PageWrapper,
    PageContent,
    ContentWrapper,
    VerticalSpacer,
    ColorBlock,
} from 'src/components/Layout';

import ProductOverview from 'src/components/ProductOverview';

import { ProductType } from 'src/types';

type Props = {
    isError: boolean;
    products: ProductType[] | null;
};
export const Shop = ({ isError, products }: Props) => {
    if (isError) {
        // TODO show error component, possibly with a retry button
    }
    return (
        <PageWrapper>
            <PageContent>
                <ColorBlock>
                    <ContentWrapper>
                        <VerticalSpacer size='extraExtraLarge' />
                        <H1>Light bulbs</H1>
                        <VerticalSpacer size='large' />
                        <Paragraph>
                            Here are some amazing light bulbs.
                        </Paragraph>
                        <VerticalSpacer size='large' />
                        {products && products.length
                            ? products.map(lightBulb => (
                                  <ProductOverview
                                      product={lightBulb}
                                      key={lightBulb.id}
                                  />
                              ))
                            : null}
                    </ContentWrapper>
                </ColorBlock>
            </PageContent>
        </PageWrapper>
    );
};

// This bit is how next.js inject data for SSR
Shop.getInitialProps = async (context: NextPageContext) => {
    let products: ProductType[] | null = null;
    let isError: boolean = false;

    try {
        // For get request
        // const url = `https://somthing.api/products/`;
        // const res = await isoFetch(url);
        // const json = await res.json();
        // product = json.results[0];
        // isError = res.status > 299 ? true : false;

        // TODO: A way to retry functions on the client if they fail in ssr hooks can't be used in ssr
        // TODO: consider pagination, totally dependent on the API
        // TODO: Do we want to store these locally or keep requesting on each page
        // incase of live changes in price and stock availability
        // Does the API cache these requests in memory until the data changes

        // this bit doesn't need to be async
        // it's all set up async for the non-existent api call
        const lightBulbs: ProductType[] | null = await data;
        products = lightBulbs || null;
    } catch (error) {
        isError = true;
    }

    return { products, isError };
};

export default Shop;
