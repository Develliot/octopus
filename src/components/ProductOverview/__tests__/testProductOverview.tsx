import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import { ProductOverview } from '../';
import { ProductType } from 'src/types';

describe('ProductOverview', () => {
    const mockProduct: ProductType = {
        id: 1,
        title: 'Energy saving light bulb',
        wattage: 25,
        packetSize: 4,
        price: 12.99,
        description:
            'Available in 7 watts, 9 watts, 11 watts Spiral Light bulb in B22, bulb switches instantly, no wait around with a warm start and flicker free features. Light up your life with this swirly wonder!',
        imageUrl: 'https://picsum.photos/500/500',
        specifications: {
            brand: 'Philips',
            weight: '77g',
            dimensions: '12.6x6.2x6.2cm',
            modelNumber: 'E27 ES',
            colour: 'Cool daylight',
        },
    };

    it('renders correctly with product', () => {
        const wrapper: ShallowWrapper = shallow(
            <ProductOverview product={mockProduct} />
        );
        expect(wrapper).toMatchSnapshot();
    });
});
