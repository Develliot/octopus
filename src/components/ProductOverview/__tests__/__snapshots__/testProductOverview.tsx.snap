// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductOverview renders correctly with product 1`] = `
<Link
  as="/product/1"
  href="/product/[id]"
  passHref={true}
>
  <styles__OverviewLink
    tabIndex={0}
    title="link to product :Energy saving light bulb"
  >
    <styles__OverviewWrapper>
      <Layout__Row>
        <styles__OverviewImage
          src="https://picsum.photos/500/500"
        />
        <styles__OverviewText>
          Energy saving light bulb
        </styles__OverviewText>
        <Memo(Price)
          price={12.99}
        />
      </Layout__Row>
    </styles__OverviewWrapper>
  </styles__OverviewLink>
</Link>
`;
